File: src/Services/Implementations/OrderService.cs
using ECommerceApi.DTOs.Orders;
using ECommerceApi.Exceptions;
using ECommerceApi.Models;
using ECommerceApi.Repositories.Interfaces;
using ECommerceApi.Services.Interfaces;

namespace ECommerceApi.Services.Implementations;

public class OrderService : IOrderService
{
    
    // Allowed order status transitions
    
    private static readonly Dictionary<OrderStatus, OrderStatus[]> AllowedTransitions = new()
    {
        [OrderStatus.Pending] = new[] { OrderStatus.Paid, OrderStatus.Cancelled },
        [OrderStatus.Paid] = new[] { OrderStatus.Shipped, OrderStatus.Cancelled },
        [OrderStatus.Shipped] = new[] { OrderStatus.Delivered },
        [OrderStatus.Delivered] = Array.Empty<OrderStatus>(),
        [OrderStatus.Cancelled] = Array.Empty<OrderStatus>()
    };

    private readonly IOrderRepository _orderRepository;
    private readonly ICartRepository _cartRepository;
    private readonly IProductRepository _productRepository;

    public OrderService(IOrderRepository orderRepository, ICartRepository cartRepository, IProductRepository productRepository)
    {
        _orderRepository = orderRepository;
        _cartRepository = cartRepository;
        _productRepository = productRepository;
    }

    /// <summary>
    /// Checkout: cart -> order. Re-validates cart contents and CURRENT
    /// stock/price before delegating the atomic write to the repository's
    /// transaction (Milestone 2 §23/§25). Never trusts a client-supplied
    /// price — totals are always derived from Product.Price read here.
    /// </summary>
    public async Task<OrderResponse> CheckoutAsync(Guid userId, Guid? shippingAddressId)
    {
        var cart = await _cartRepository.FindOrCreateByUserAsync(userId);

        if (cart.Items.Count == 0)
        {
            throw new ValidationException("Cannot checkout with an empty cart");
        }

        decimal totalAmount = 0m;
        var orderItems = new List<CreateOrderItemInput>();

        foreach (var cartItem in cart.Items)
        {
            var product = await _productRepository.FindByIdAsync(cartItem.ProductId);
            if (product is null)
            {
                // Covers both "never existed" and "since soft-deleted"
                throw new NotFoundAppException($"Product {cartItem.ProductId}");
            }
            if (cartItem.Quantity > product.StockQuantity)
            {
                throw new ValidationException($"\"{product.Name}\" only has {product.StockQuantity} unit(s) in stock");
            }

            totalAmount += product.Price * cartItem.Quantity;
            orderItems.Add(new CreateOrderItemInput(product.Id, cartItem.Quantity, product.Price));
        }

        var order = await _orderRepository.CreateFromCartAsync(
            new CreateOrderInput(userId, shippingAddressId, Math.Round(totalAmount, 2), orderItems));

        return Map(order);
    }

    public async Task<OrderResponse> GetOrderForUserAsync(Guid orderId, Guid requesterId, Role requesterRole)
    {
        var order = await _orderRepository.FindByIdAsync(orderId);
        if (order is null)
        {
            throw new NotFoundAppException("Order");
        }

        // Ownership check: role alone is insufficient
        if (requesterRole != Role.Admin && order.UserId != requesterId)
        {
            throw new ForbiddenAppException("You do not have access to this order");
        }

        return Map(order);
    }

    public async Task<PagedOrdersResponse> ListOrdersForUserAsync(Guid userId, int page, int limit)
    {
        var result = await _orderRepository.FindByUserAsync(userId, page, limit);
        return new PagedOrdersResponse { Items = result.Items.Select(Map).ToList(), Total = result.Total };
    }

    public async Task<PagedOrdersResponse> ListAllOrdersAsync(int page, int limit, OrderStatus? status)
    {
        var result = await _orderRepository.FindAllAsync(page, limit, status);
        return new PagedOrdersResponse { Items = result.Items.Select(Map).ToList(), Total = result.Total };
    }

    public async Task<OrderResponse> UpdateStatusAsync(Guid orderId, OrderStatus nextStatus)
    {
        var order = await _orderRepository.FindByIdAsync(orderId);
        if (order is null)
        {
            throw new NotFoundAppException("Order");
        }

        var allowed = AllowedTransitions[order.Status];
        if (!allowed.Contains(nextStatus))
        {
            throw new ValidationException($"Cannot transition order from '{order.Status}' to '{nextStatus}'");
        }

        var updated = await _orderRepository.UpdateStatusAsync(orderId, nextStatus);
        return Map(updated);
    }

    private static OrderResponse Map(Order order)
    {
        return new OrderResponse
        {
            Id = order.Id,
            UserId = order.UserId,
            Status = order.Status.ToString().ToLowerInvariant(),
            TotalAmount = order.TotalAmount,
            ShippingAddressId = order.ShippingAddressId,
            CreatedAt = order.CreatedAt,
            Items = order.Items.Select(i => new OrderItemResponse
            {
                Id = i.Id,
                ProductId = i.ProductId,
                ProductName = i.Product?.Name ?? "Unknown product",
                Quantity = i.Quantity,
                UnitPriceAtPurchase = i.UnitPriceAtPurchase
            }).ToList()
        };
    }
}

File: legacy-nodejs/src/services/order.service.ts
import { ICartRepository, IOrderRepository, IProductRepository } from '../repositories/interfaces';
import { OrderRecord, OrderStatus } from '../types/domain';
import { Role } from '../types/auth';
import { ForbiddenError, NotFoundError, ValidationError } from '../errors/AppError';

/**
 * Allowed order status transitions (Milestone 1 §10 / Milestone 2 §26).
 * A transition is valid only if `to` appears in `ALLOWED_TRANSITIONS[from]`.
 */
const ALLOWED_TRANSITIONS: Record<OrderStatus, OrderStatus[]> = {
  pending: ['paid', 'cancelled'],
  paid: ['shipped', 'cancelled'],
  shipped: ['delivered'],
  delivered: [],
  cancelled: []
};

export class OrderService {
  constructor(
    private readonly orderRepository: IOrderRepository,
    private readonly cartRepository: ICartRepository,
    private readonly productRepository: IProductRepository
  ) {}

  /**
   * Checkout: cart -> order. Re-validates cart contents and current stock
   * before delegating the actual atomic write (order + items + stock
   * decrement + cart clear) to the repository's transaction
   * (Milestone 2 §25). This method contains the *business rules*; the
   * repository contains the *atomicity mechanism*.
   */
  async checkout(userId: string, shippingAddressId: string | null): Promise<OrderRecord> {
    const cart = await this.cartRepository.findOrCreateByUser(userId);

    if (cart.items.length === 0) {
      throw new ValidationError('Cannot checkout with an empty cart');
    }

    let totalAmount = 0;
    const orderItems: Array<{ productId: string; quantity: number; unitPriceAtPurchase: number }> = [];

    for (const cartItem of cart.items) {
      const product = await this.productRepository.findById(cartItem.productId);
      if (!product || !product.isActive) {
        throw new NotFoundError(`Product ${cartItem.productId}`);
      }
      if (cartItem.quantity > product.stockQuantity) {
        throw new ValidationError(
          `"${product.name}" only has ${product.stockQuantity} unit(s) in stock`
        );
      }
      totalAmount += product.price * cartItem.quantity;
      orderItems.push({
        productId: product.id,
        quantity: cartItem.quantity,
        unitPriceAtPurchase: product.price
      });
    }

    return this.orderRepository.createFromCart({
      userId,
      shippingAddressId,
      totalAmount: Math.round(totalAmount * 100) / 100,
      items: orderItems
    });
  }

  async getOrderForUser(orderId: string, requesterId: string, requesterRole: Role): Promise<OrderRecord> {
    const order = await this.orderRepository.findById(orderId);
    if (!order) {
      throw new NotFoundError('Order');
    }
    // Ownership check (Milestone 1 §17): role alone is insufficient — a
    // customer may only view their own order, regardless of authentication.
    if (requesterRole !== 'admin' && order.userId !== requesterId) {
      throw new ForbiddenError('You do not have access to this order');
    }
    return order;
  }

  async listOrdersForUser(userId: string, page: number, limit: number) {
    return this.orderRepository.findByUser(userId, page, limit);
  }

  async listAllOrders(page: number, limit: number, status?: OrderStatus) {
    return this.orderRepository.findAll(page, limit, status);
  }

  async updateStatus(orderId: string, nextStatus: OrderStatus): Promise<OrderRecord> {
    const order = await this.orderRepository.findById(orderId);
    if (!order) {
      throw new NotFoundError('Order');
    }

    const allowed = ALLOWED_TRANSITIONS[order.status];
    if (!allowed.includes(nextStatus)) {
      throw new ValidationError(`Cannot transition order from '${order.status}' to '${nextStatus}'`);
    }

    return this.orderRepository.updateStatus(orderId, nextStatus);
  }
}
